### ISSUE TITLE
Issue with VectorTools::compute_no_normal_flux_constraints

### ISSUE THREAD

**Issue by sairajat** [dealii/dealii#3323]
Hello all,

I am using VectorTools::compute_no_normal_flux_constraints() to apply some constraints of the form X.n = 0, where x and n are 3 dimensional vectors.

Now, upon applying this constraint, I find that for a particular case, the code hangs inside this function.
Upon further digging, I discover that it is stuck in a for loop in line 545 in
https://github.com/dealii/dealii/blob/master/source/grid/tria_boundary.cc#L545

I have a couple of questions here.

1. The simple test case to demonstrate the problem works fine with no issue but the original code doesn't. 
I think it is because of numerical error while i/o (writing and then reading from file). 

My mesh is not distorted but it still is stuck inside that loop and the Assert is triggered upon running with debug mode. I am attaching pictures of my mesh at the moment of calling that function.

![back](https://cloud.githubusercontent.com/assets/2487250/19845013/2626092e-9f09-11e6-91df-e55fb9b57c4c.png)
![front](https://cloud.githubusercontent.com/assets/2487250/19845014/26263d9a-9f09-11e6-9ccf-c03a0e0b2cd3.png)

I think it is happening because the elements in the z direction are long. So, in 10 iterations, due to numerical inaccuracies, sometime (like in 52nd iteration) the required tolerance is not achieved in 10 iterations. 

2. Also, once I got to know of the problem, I played with the mesh right. Right before it went into the compute_no_normal_flux constraints function, I changed the thickness of my mesh and made it very thin. So, this time, everything remains the same but only the mesh changes (doesn't make much sense physically). Surprisingly, the code made it through without hanging this time. This corroborates my earlier claim that this may be happening because of the large aspect ratio in the z direction. 

3. Should we just get stuck in the code in case the mesh is severely distorted or like the one in my case while running in release mode. Shouldn't there be a better option to decide what to do in a situation like this  ? 

4. Why are we using non-linear least squares ? Can we just not use Newton Raphson Technique to find the point in the reference coordinate ? Is the mapping not unique that we need to find the minimum of the distance rather than just using Newton Raphson to find the point itself (approximately)?

5. Also, for the stopping criteria, which one of these two should be used the delta_xi.norm() < eps or J.norm() < eps ? Currently, delta_xi.norm() < eps is used with eps = 1e^(-12)

**Comment by spco** [comment]
In my experience, there are a number of locations in the library where the Release version will loop endlessly because of an Assert not being triggered. While this can be frustrating, the design aim has always been that users run in Release mode _only_ after ensuring that Debug runs without failed Asserts. So on (2) (the only one I can really comment on!) I don't think there is a better option. Just test your mesh in Debug mode before switching to Release. I appreciate that this is time-consuming, particularly when running a big 3D simulation which takes forever in Debug mode. 

If anyone else has a better strategy for such cases I'd also love to hear it!

**Comment by drwells** [comment]
This is a good point. We should probably use a relative and not absolute tolerance (the error probably scales with the cell area).

I don't know enough about this mapping to determine why the author did not use Newton-Raphson. At a glance one should be able to guess a normal vector by using three out of four points and then using Newton iteration to refine things. This is worth a closer look.

**Comment by bangerth** [comment]
I think the error is computed in reference coordinates, where relative = absolute tolerance.

The issue is more likely the ill-conditionedness of the problem. Instead of running a Newton iteration on the actual Newton system, I think it would be useful to scale the problem (i.e., precondition it) with a matrix that encodes the dimensions in the various coordinate directions, possibly rotated into the coordinate system of the cell. I bet that can be done by just multiplying everything by the inverse of the Jacobian of the mapping at the center of the cell.

**Comment by sairajat** [comment]
Thank you all for your prompt reply.
 I also see that it is due to ill-conditioning of the matrix H begetting numerical inaccuracies. 

I would also love to hear some comments on points 3 and 4 above. I am curious to know as to why (if) Gauss-Newton method is the preferred choice over Newton Raphson ?

Also, through this discussion, I would like to learn what other strategies could be implemented / exist (in deal.ii or other softwares) that do not cause hanging of the code when such a thing is encountered ?

Simple solution could be just using a break / AssertThrow statement. Are there some other techniques as well ?

**Comment by sairajat** [comment]
Also, why do we even want to map the points from the real cell to the natural coordinates. If we map the nodes on the  real cell, will it not always map to the nodes on the natural coordinates ? We can avoid using the Newton-Gauss method all together. We can just use the orientation and then vertex number  [0 - 7] to determine the coordinates in the natural system. Is that not right ? Am I missing something here ?

**Comment by bangerth** [comment]
@sairajat -- would you mind creating a small test program that demonstrates the issue? It doesn't have to do anything useful, all that it needs to do is (i) create a mesh and dofhandler, (ii) call `VectorTools::compute_no_normal_flux_constraints()` to demonstrate that it hangs.

**Comment by bangerth** [comment]
About your other questions:
3/ Of course, getting stuck in the iteration is the wrong thing. If we can't converge, and in the absence of a better algorithm, the the right thing is to terminate the program because it can't do what the user wanted. (`AssertThrow` is entirely appropriate.)

4/ I suspect that using Newton-Raphson on the nonlinear equation would work just as well as using Gauss-Newton on the least-squares problem. One could easily try this out.

About why we need this to begin with: It is correct that (unless one uses an Eulerian mapping) the vertices of the cell are mapped to the vertices of the reference cell. I suspect, though, that the function does not just map vertices -- but without a backtrace, I can't tell where we actually call this function from. That's why a testcase would be so useful :-)

**Comment by sairajat** [comment]
@bangerth  Prof.  Bangerth, I was trying to make a test that demonstrates it. While I try that, I have found something that I think might be wrong.

I am moving the mesh after every time step like described in the post at
[google dealii forum link](https://groups.google.com/forum/#!searchin/dealii/move$20mesh|sort:relevance/dealii/8I21UqIWAmU/yuJT877WAAAJ)

Here, only `locally_owned_elements` are moved which is similar to step 18. We only move `locally_owned_cells` while mesh movement. Right ?

While applying boundary conditions, we take average of normals from adjacent cells and then do everything with it. My question is: Shouldn't this averaging be done on only `locally_owned_cells` and not `other cells (ghost and artificial)`  as not moving `ghost cells` might make them distorted ?

This is why I think averaging should only be done on `locally_owned_cells`. But in file [Vector tools templates line 5713](https://github.com/dealii/dealii/blob/master/include/deal.II/numerics/vector_tools.templates.h#L5713), averaging is done for all cells: `!cell->is_artificial` which includes `locally_owned_elements` and `ghost_elements`. Hence, the distortion can come in the picture, even if the mesh looks nice (we only see `locally_owned_elements`).

**Comment by sairajat** [comment]
@bangerth  Respected Prof. Bangerth, I am unable to reproduce the issue with just using a rectangular mesh with large aspect ratio. I suspect that this is because of reason I mentioned in my last comment.

**Comment by bangerth** [comment]
I think there's not very much we can do without a way to reproduce the issue. It's not important that it is a simple rectangular mesh, though. Any testcase would help. For the moment, let me close the issue (since there's not much we can do for now), but feel free to reopen if you do have a testcase that allows us to reproduce things.

Regarding the mesh movement -- I think you need to move *all* cells, even the ones not locally owned or ghosted. For sure you need to move ghost cells because they are important for the local computations, including for what `compute_no_normal_flux_constraints()` does: it does look to neighboring elements as well, as described in the function's documentation. If you plan to do mesh refinement and coarsening, you will also have to make sure that artificial cells are moved as appropriate. This may not be easy to achieve, though.

**Comment by sairajat** [comment]
@bangerth  Hello Professor, I have found the issue.
I was unaware that the ghost cells needs to be moved as well. I posted this doubt in the forum and got my answer. 

So, the problem was the the ghost cells were not being moved and hence mesh (part which is never visible in paraview) was becoming distorted.

The solution is when moving the mesh, move the ghost cell as well: either yourself (using loop as in step 18 but for all cells that are not artificial) or use the internal function [communicate_locally_moved_vertices](https://www.dealii.org/8.4.1/doxygen/deal.II/classparallel_1_1distributed_1_1Triangulation.html#a247598f1323a9f847832e60d6c840469)

Thanks a lot for your previous comments and help.
